plugins/modules/redfish_firmware_rollback.py

```diff
@@ -0,0 +1,360 @@
+
+
+def require_session(idrac, module):
+    session_id, token = "", None
+    payload = {'UserName': module.params["username"], 'Password': module.params["password"]}
+    path = SESSION_RESOURCE_COLLECTION["SESSION"]
+    resp = idrac.invoke_request('POST', path, data=payload, api_timeout=120)
+    if resp and resp.success:
+        session_id = resp.json_data.get("Id")
+        token = resp.headers.get('X-Auth-Token')
+    return session_id, token
+
+
+def wait_for_redfish_idrac_reset(module, redfish_obj, wait_time_sec, interval=30):
+    time.sleep(interval // 2)
+    msg = RESET_UNTRACK
+    wait = wait_time_sec
+    track_failed = True
+    resetting = False
+    while wait > 0 and track_failed:
+        try:
+            redfish_obj.invoke_request("GET", MANAGERS_URI, api_timeout=120)
+            msg = RESET_SUCCESS
+            track_failed = False
+            break
+        except HTTPError as err:
+            if err.getcode() == 401:
+                new_redfish_obj = Redfish(module.params, req_session=True)
+                sid, token = require_session(new_redfish_obj, module)
+                redfish_obj.session_id = sid
+                redfish_obj._headers.update({"X-Auth-Token": token})
+                track_failed = False
+                if not resetting:
+                    resetting = True
+                break
+            time.sleep(interval)
+            wait = wait - interval
+            resetting = True
+        except URLError:
+            time.sleep(interval)
+            wait = wait - interval
+            if not resetting:
+                resetting = True
+        except Exception:
+            time.sleep(interval)
+            wait = wait - interval
+            resetting = True
+    return track_failed, resetting, msg
+
+
+def simple_update(redfish_obj, preview_uri, update_uri):
+    job_ids = []
+    for uri in preview_uri:
+        resp = redfish_obj.invoke_request("POST", update_uri, data={"ImageURI": uri})
+        time.sleep(30)
+        task_uri = resp.headers.get("Location")
+        task_id = task_uri.split("/")[-1]
+        job_ids.append(task_id)
+    return job_ids
+
+
+def rollback_firmware(redfish_obj, module, preview_uri, reboot_uri, update_uri):
+    current_job_status, failed_cnt, resetting = [], 0, False
+    job_ids = simple_update(redfish_obj, preview_uri, update_uri)
+    if module.params["reboot"] and preview_uri:
+        payload = {"ResetType": "ForceRestart"}
+        job_resp_status, reset_status, reset_fail = wait_for_redfish_reboot_job(redfish_obj, SYSTEM_RESOURCE_ID,
+                                                                                payload=payload)
+        if reset_status and job_resp_status:
+            job_uri = MANAGER_JOB_ID_URI.format(job_resp_status["Id"])
+            job_resp, job_msg = wait_for_redfish_job_complete(redfish_obj, job_uri)
+            job_status = job_resp.json_data
+            if job_status["JobState"] != "RebootCompleted" and job_msg:
```

Comment:
line 301 to 303, can be changed to

`if job_status["JobState"] != "RebootCompleted:
     if job_msg:
       do this
     else:
       do this.
`

Here we are checking JobState twice in line 301 and 303.

Reply:
Yes will update this